certbot-apache/certbot_apache/augeasparser.py

```diff
@@ -67,8 +67,30 @@ def __eq__(self, other):
         return False
 
     def set_parameters(self, parameters):
-        """Sets the parameters for DirectiveNode"""
-        self.parameters = parameters
+        """
+        Sets parameters of a DirectiveNode or BlockNode object.
+
+        :param list parameters: List of all parameters for the node to set.
+        """
+        orig_params = self._aug_get_params(self.metadata["augeaspath"])
+
+        # Clear out old parameters
+        for _ in orig_params:
+            # When the first parameter is removed, the indices get updated
+            param_path = "{}/arg[1]".format(self.metadata["augeaspath"])
+            self.parser.aug.remove(param_path)
+        # Insert new ones
+        for pi, _ in enumerate(parameters):
+            param_path = "{}/arg[{}]".format(self.metadata["augeaspath"], pi+1)
+            self.parser.aug.set(param_path, parameters[pi])
+
+        self.parameters = tuple(parameters)
```

Comment:
1. Good catch. I'm not sure about this. I'll look into some real world examples.
2. As you noted, having `parameters` as property would mitigate the out-of-sync issue and would probably work pretty well. I think it was initially a property but we decided to move away from that at some point of the interface design, for a reason I can't remember right now. I'll look into it.

Reply:
> If parameters contains a variable, it will not be expanded unlike functions like parser.get_arg. Is this the behavior we want?

I looked this up. The code in `parser.get_arg` is actually introducing a bug. Augeas expands the variables itself, and if the code in that function replaces something, it actually ends up introducing a bug instead.

Reply:
That sounds like the right approach to me. I think we should make sure that this is done the same way in the `apacheconfig` `ParserNode`.